ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml

```diff
@@ -0,0 +1,11 @@
+chain:
+  as: cucushift-installer-rehearse-aws-upi-deprovision
+  steps:
+  - chain: gather
+  - ref: upi-deprovision-aws-stage1
+  - ref: ipi-deprovision-deprovision
+  - ref: upi-deprovision-aws-stage2
```

Comment:
I don't understand the need to split `stage1` and `stage2` with the IPI deprovisioner in the middle.  Our [uninstall docs][1] have no section for clusters built on user-provisioned infrastructure.  It seems like I commited a `delete-stack` , `destroy cluster`, `delete-stack`-again pattern in 602791d72c475d5878676305eca1b27274cffdef, although I neglected to record my motivation in the commit message.  Maybe it was:

1. Handwave, maybe stack deletion just works.  Try it.
2. Maybe it didn't work, because some cluster-created resource is blocking things?  Try a `destroy cluster`, in case there is anything cluster-created (or tagged as cluster-owned in a stack) that the installer can remove.
3. Hopefully that got any blockers out of the way.  Try the stacks again.

Depending on how solidly user-provisioned the workflow is intending to be, 2 and 3 may be unecessary complication.

[1]: https://docs.openshift.com/container-platform/4.11/installing/installing_aws/uninstalling-cluster-aws.html

Reply:
Thank you @wking for providing so detailed info, I will try and make the deprovision steps as easy as possible.

Reply:
@wking The `*.apps.[cluster-name].[base-domainname]` record created by ingress will cause deleting stack `02_cluster_infra.yaml` failed, that's why I put it in  `stage2`,  after the `destroy` command. 

Other resources, like s3 buckets created image registry, IAM policies created by CR, also need to be deleted by `openshift-install destroy`. 

I think that's why you used `delete-stack` , `destroy cluster`, `delete-stack`-again pattern in https://github.com/openshift/release/commit/602791d72c475d5878676305eca1b27274cffdef, also the similar process is being used in [existing UPI post action](https://github.com/openshift/release/blob/master/ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml#L5-L7).

To make the process easy to understand, I will use the similar one:
1. delete all stacks (deleting infrastructure stack will fail)
2. `openshift-install destroy` 
3. delete failed (or all)  stacks again.

Reply:
I reused existing `upi-deprovision-aws` and `upi-deprovision-aws-check-destroy` steps, and removed `stage[12]`.

Reply:
ok, pinning this down so I don't forget:
1. First stack deletion.  It won't complete (blocked by cluster-created resources like ingress DNS records, etc.).  But we need to terminate instances and otherwise hobble the cluster as much as we can, so that `openshift-install destroy` will be able to delete cluster-created resources.  If we opened with `openshift-install destroy`, the cluster might remain healthy enough to create replacement resources.
2. `openshift-install destroy` to remove cluster-created resources, which the stacks don't know about.
3. Remove the stacks, now that they're unblocked.